crates/codec/src/api/wasm/buffer.rs

```diff
@@ -0,0 +1,258 @@
+use std::convert::{TryFrom, TryInto};
+use std::ops::{Range, RangeFrom};
+
+/// ## WASM Buffer Layout
+///
+/// Each WASM Buffer contains 2 section: `Header` and `Data`
+///
+/// +--------------------------------+
+/// | Header Section |  Data Section |
+/// +--------------------------------+
+///
+///
+/// ### WASM Buffer Header Section
+///
+/// Each Buffer is prefixed with `Header` consisting of 8 bytes.
+///
+/// The first 4 bytes are the byte-length of the buffer.
+/// The remaining 4 bytes are the capacity byte-length of the buffer.
+///
+/// The reason we need both `length` and `capacity` and due to the implementation
+/// of Rust `Vec`. Even though we use `Vec::with_capacity` we still prefer to store
+/// explicitly in the `Header` the `capacity` returned by `Vec#into_raw_parts`.
+///
+/// See also `Vec#reserve_exact` documentation:
+///
+/// ```md
+/// Note that the allocator may give the collection more space than it
+/// requests. Therefore, capacity can not be relied upon to be precisely
+/// minimal. Prefer `reserve` if future insertions are expected.
+/// ```
+///
+/// #### WASM Buffer Header Layout
+///
+/// +------------------+--------------------+
+/// | length (4 bytes) | capacity (4 bytes) |
+/// +------------------+--------------------+
+///
+/// Both `length` and `capacity` are laid out in Big-Endian order
+///
+///
+/// ## WASM Buffer Data Section
+///
+/// Contains the raw data of the buffer.
+///
+/// Allocates a new WASM buffer having `Data` of `length` bytes.
+///
+/// The total allocation size of the buffer will always be bigger due to the
+/// `Header` section.
+/// If for the `capacity` of the `Data` will be bigger - it will also increase
+/// the amount of allocated data.
+#[derive(Debug, PartialEq)]
+pub struct Buffer(Option<Vec<u8>>, bool);
+
+impl Buffer {
+    const LEN: Range<usize> = 0..4;
+    const CAPACITY: Range<usize> = 4..8;
+    const DATA: RangeFrom<usize> = 8..;
+
+    const HEADER_SIZE: usize = 8;
+
+    const OK_MARKER: u8 = 1;
+    const ERR_MARKER: u8 = 0;
+
+    pub fn alloc(len: u32) -> Self {
+        let mut vec = vec![0; len as usize + Self::HEADER_SIZE];
+        vec.shrink_to_fit();
+
+        assert_eq!(vec.capacity(), len as usize + Self::HEADER_SIZE);
+        assert_eq!(vec.len(), len as usize + Self::HEADER_SIZE);
+
+        let mut buf = Self(Some(vec), true);
+
+        buf.set_capacity(len);
+        buf.set_len(len);
+
+        buf
+    }
+
+    #[cfg(test)]
+    pub fn then_free(mut self) -> Self {
```

Comment:
Why not name it `free`?

Reply:
Because it doesn't free `Buffer`. `Buffer` has two modes - either leaking (default, given that it's what you usually want when doing FFI) or freeing. This changes the memory mode, but it doesn't immediately free. Memory will only get freed on drop.